crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2830,6 +2830,93 @@ impl<'db> TypeInferenceBuilder<'db> {
+                // type.
+                let bool_type = if let Symbol::Type(bool_method, _) =
+                    class.class_member(self.db, "__bool__")
+                {
+                    let call = bool_method.call(self.db, &[operand_type]);
+                    let bool_rt = match call.return_ty_result(
+                        self.db,
+                        AnyNodeRef::ExprUnaryOp(unary),
+                        &mut self.diagnostics,
+                    ) {
+                        Ok(t) => t,
+                        Err(_) => Type::BooleanLiteral(true),
+                    };
+
+                    if bool_rt.is_subtype_of(
+                        self.db,
+                        KnownClass::Bool.to_class(self.db).to_instance(self.db),
+                    ) {
+                        bool_rt.bool(self.db)
+                    } else {
+                        self.diagnostics.add(
+                            unary.into(),
+                            "unsupported-operator",
+                            format_args!(
+                                "Method __bool__ for type `{}` should return `bool`, returned type `{}`",
+                                operand_type.display(self.db),
+                                bool_rt.display(self.db)
+                            ),
+                        );
+                        Truthiness::Ambiguous
+                    }
+                } else {
+                    Truthiness::AlwaysTrue
+                };
+
+                let len_type = if let Symbol::Type(len_method, _) =
+                    class.class_member(self.db, "__len__")
+                {
+                    let call = len_method.call(self.db, &[operand_type]);
+                    let len_rt = match call.return_ty_result(
+                        self.db,
+                        AnyNodeRef::ExprUnaryOp(unary),
+                        &mut self.diagnostics,
+                    ) {
+                        Ok(t) => t,
+                        Err(_) => Type::BooleanLiteral(true),
+                    };
+                    // TODO: len should be an integer >= 0 and should fit into sys.maxsize
+                    // https://docs.python.org/3/reference/datamodel.html#object.__len__
+                    // The type returned by len must be able to be interpreted as integer
+
+                    if len_rt.is_subtype_of(
+                        self.db,
+                        KnownClass::Int.to_class(self.db).to_instance(self.db),
+                    ) {
+                        len_rt.bool(self.db)
+                    } else {
+                        self.diagnostics.add(
+                            unary.into(),
+                            "unsupported-operator",
+                            format_args!(
+                                "Method __len__ for type `{}` should return `int`, returned type `{}`",
```

Comment:
Pyright is okay with this:
https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2AUNQMaXkDOTUAggFzVQ9QAmJYFAD6wyiRSiAFExKVgASgC0APijBKYcjC689UECRgBXECigAGAHQXahgG4kqw%2BAhJSUYfGykKFQA

But python itself is not:
```
    not A()
TypeError: 'float' object cannot be interpreted as an integer
```

So I decided to add this.

Reply:
I agree that we should have the other diagnostic, but I think it should not happen in this PR for two different reasons:

1) I don't this PR actually should involve `__len__` at all.
2) I think the diagnostic should be on the class definition, not at the usage site.